File: plugins/expo-autofill-plugin/ios-template/PearPassAutofillExtension/AuthenticatorDataBuilder.swift
//
//  AuthenticatorDataBuilder.swift
//  PearPassAutoFillExtension
//
//  Builds WebAuthn authenticator data structures
//  Compatible with browser extension's buildAuthenticatorData.js
//

import Foundation
import CryptoKit

struct AuthenticatorDataBuilder {

    // MARK: - Flags

    /// WebAuthn authenticator data flags
    struct Flags: OptionSet {
        let rawValue: UInt8

        /// User Present (UP) - Bit 0
        static let userPresent = Flags(rawValue: 1 << 0)
        /// User Verified (UV) - Bit 2
        static let userVerified = Flags(rawValue: 1 << 2)
        /// Backup Eligibility (BE) - Bit 3
        static let backupEligible = Flags(rawValue: 1 << 3)
        /// Backup State (BS) - Bit 4
        static let backupState = Flags(rawValue: 1 << 4)
        /// Attested Credential Data (AT) - Bit 6
        static let attestedCredentialData = Flags(rawValue: 1 << 6)
        /// Extension Data (ED) - Bit 7
        static let extensionData = Flags(rawValue: 1 << 7)

        /// Standard flags for registration: UP + UV + BE + BS + AT = 0x5D
        /// Matches browser extension exactly
        static let registration: Flags = [
            .userPresent,
            .userVerified,
            .backupEligible,
            .backupState,
            .attestedCredentialData
        ]

        /// Standard flags for assertion: UP + UV + BE + BS = 0x1D
        static let assertion: Flags = [
            .userPresent,
            .userVerified,
            .backupEligible,
            .backupState
        ]
    }

    // MARK: - Constants

    /// Software authenticator AAGUID (all zeros)
    /// Matches browser extension's software authenticator
    static let softwareAAGUID = Data(repeating: 0, count: 16)

    // MARK: - Build Authenticator Data for Registration

    /// Build authenticator data for passkey registration
    /// Structure: rpIdHash (32) || flags (1) || signCount (4) || aaguid (16) || credIdLen (2) || credId || coseKey
    /// - Parameters:
    ///   - rpId: Relying party identifier (domain)
    ///   - credentialId: Credential ID bytes
    ///   - publicKey: P256 signing public key
    /// - Returns: Complete authenticator data bytes
    static func buildForRegistration(
        rpId: String,
        credentialId: Data,
        publicKey: P256.Signing.PublicKey
    ) -> Data {
        var authData = Data()

        // 1. RP ID Hash (32 bytes) - SHA-256 of rpId
        let rpIdHash = PasskeyCrypto.sha256(rpId)
        authData.append(rpIdHash)

        // 2. Flags (1 byte) - 0x5D for registration
        authData.append(Flags.registration.rawValue)

        // 3. Sign count (4 bytes, big-endian) - start at 0
        authData.append(contentsOf: [0x00, 0x00, 0x00, 0x00])

        // 4. AAGUID (16 bytes) - all zeros for software authenticator
        authData.append(softwareAAGUID)

        // 5. Credential ID length (2 bytes, big-endian)
        let credIdLength = UInt16(credentialId.count)
        authData.append(contentsOf: withUnsafeBytes(of: credIdLength.bigEndian) { Array($0) })

        // 6. Credential ID
        authData.append(credentialId)

        // 7. COSE public key (CBOR-encoded)
        let coseKey = encodeCOSEPublicKey(publicKey)
        authData.append(coseKey)

        return authData
    }

    // MARK: - Build Authenticator Data for Assertion

    /// Build authenticator data for passkey assertion
    /// Structure: rpIdHash (32) || flags (1) || signCount (4)
    /// - Parameters:
    ///   - rpId: Relying party identifier (domain)
    ///   - signCount: Current signature counter value (default 0)
    /// - Returns: Authenticator data bytes for assertion (37 bytes)
    static func buildForAssertion(
        rpId: String,
        signCount: UInt32 = 0
    ) -> Data {
        var authData = Data()

        // 1. RP ID Hash (32 bytes) - SHA-256 of rpId
        let rpIdHash = PasskeyCrypto.sha256(rpId)
        authData.append(rpIdHash)

        // 2. Flags (1 byte) - 0x1D for assertion
        authData.append(Flags.assertion.rawValue)

        // 3. Sign count (4 bytes, big-endian)
        authData.append(contentsOf: withUnsafeBytes(of: signCount.bigEndian) { Array($0) })

        return authData
    }

    // MARK: - COSE Key Encoding

    /// Encode P256 public key as COSE_Key (CBOR)
    /// Structure: {1: 2, 3: -7, -1: 1, -2: x, -3: y}
    /// - kty (1): EC2 (2)
    /// - alg (3): ES256 (-7)
    /// - crv (-1): P-256 (1)
    /// - x (-2): x-coordinate (32 bytes)
    /// - y (-3): y-coordinate (32 bytes)
    /// - Parameter publicKey: P256 signing public key
    /// - Returns: CBOR-encoded COSE key
    private static func encodeCOSEPublicKey(_ publicKey: P256.Signing.PublicKey) -> Data {
        // Get raw X9.63 representation: 04 || x (32 bytes) || y (32 bytes)
        let rawKey = publicKey.x963Representation
        let xCoord = rawKey[1..<33]
        let yCoord = rawKey[33..<65]

        var coseKey = Data()

        // Map with 5 items
        coseKey.append(CBOREncoder.startMap(count: 5))

        // 1 (kty): 2 (EC2)
        coseKey.append(CBOREncoder.encodeUInt(1))
        coseKey.append(CBOREncoder.encodeUInt(2))

        // 3 (alg): -7 (ES256) -> encode as negative: -7 = -1 - 6
        coseKey.append(CBOREncoder.encodeUInt(3))
        coseKey.append(CBOREncoder.encodeNegInt(6))

        // -1 (crv): 1 (P-256) -> -1 = -1 - 0
        coseKey.append(CBOREncoder.encodeNegInt(0))
        coseKey.append(CBOREncoder.encodeUInt(1))

        // -2 (x): x-coordinate -> -2 = -1 - 1
        coseKey.append(CBOREncoder.encodeNegInt(1))
        coseKey.append(CBOREncoder.encodeBytes(Data(xCoord)))

        // -3 (y): y-coordinate -> -3 = -1 - 2
        coseKey.append(CBOREncoder.encodeNegInt(2))
        coseKey.append(CBOREncoder.encodeBytes(Data(yCoord)))

        return coseKey
    }

    // MARK: - Attestation Object

    /// Encode attestation object (CBOR map)
    /// Structure: {"fmt": "none", "attStmt": {}, "authData": bytes}
    /// - Parameter authData: Authenticator data bytes
    /// - Returns: CBOR-encoded attestation object
    static func encodeAttestationObject(authData: Data) -> Data {
        var attestationObject = Data()

        // Map with 3 items
        attestationObject.append(CBOREncoder.startMap(count: 3))

        // "fmt": "none" - no attestation
        attestationObject.append(CBOREncoder.encodeText("fmt"))
        attestationObject.append(CBOREncoder.encodeText("none"))

        // "attStmt": {} - empty attestation statement
        attestationObject.append(CBOREncoder.encodeText("attStmt"))
        attestationObject.append(CBOREncoder.startMap(count: 0))

        // "authData": bytes
        attestationObject.append(CBOREncoder.encodeText("authData"))
        attestationObject.append(CBOREncoder.encodeBytes(authData))

        return attestationObject
    }

    // MARK: - Client Data JSON

    /// Build client data JSON for registration
    /// - Parameters:
    ///   - challenge: Challenge bytes
    ///   - origin: Origin URL string
    /// - Returns: Client data JSON as Data
    static func buildClientDataJSONForRegistration(
        challenge: Data,
        origin: String
    ) -> Data {
        let clientData: [String: Any] = [
            "type": "webauthn.create",
            "challenge": challenge.base64URLEncodedString(),
            "origin": origin,
            "crossOrigin": false
        ]

        // Use sorted keys for consistent output
        let options = JSONSerialization.WritingOptions.sortedKeys
        return (try? JSONSerialization.data(withJSONObject: clientData, options: options)) ?? Data()
    }

    /// Build client data JSON for assertion
    /// - Parameters:
    ///   - challenge: Challenge bytes
    ///   - origin: Origin URL string
    /// - Returns: Client data JSON as Data
    static func buildClientDataJSONForAssertion(
        challenge: Data,
        origin: String
    ) -> Data {
        let clientData: [String: Any] = [
            "type": "webauthn.get",
            "challenge": challenge.base64URLEncodedString(),
            "origin": origin,
            "crossOrigin": false
        ]

        let options = JSONSerialization.WritingOptions.sortedKeys
        return (try? JSONSerialization.data(withJSONObject: clientData, options: options)) ?? Data()
    }
}

File: plugins/expo-autofill-plugin/android-template/java/autofill/crypto/AuthenticatorDataBuilder.java
package com.pears.pass.autofill.crypto;

import java.io.ByteArrayOutputStream;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.PublicKey;
import java.security.interfaces.ECPublicKey;

/**
 * Builds WebAuthn authenticator data structures.
 * Compatible with browser extension's buildAuthenticatorData.js
 *
 * Authenticator Data Flags:
 *   UP (User Present)           - Bit 0: 0x01
 *   UV (User Verified)          - Bit 2: 0x04
 *   BE (Backup Eligibility)     - Bit 3: 0x08
 *   BS (Backup State)           - Bit 4: 0x10
 *   AT (Attested Credential)    - Bit 6: 0x40
 *   ED (Extension Data)         - Bit 7: 0x80
 *
 *   Registration flags: UP + UV + BE + BS + AT = 0x5D
 *   Assertion flags:    UP + UV + BE + BS      = 0x1D
 */
public class AuthenticatorDataBuilder {

    private static final byte FLAGS_REGISTRATION = 0x5D;
    private static final byte FLAGS_ASSERTION = 0x1D;

    /** Software authenticator AAGUID (all zeros, 16 bytes) */
    private static final byte[] SOFTWARE_AAGUID = new byte[16];

    /**
     * Build authenticator data for passkey registration.
     * Structure: rpIdHash(32) || flags(1) || signCount(4) || aaguid(16) || credIdLen(2) || credId || coseKey
     */
    public static byte[] buildForRegistration(String rpId, byte[] credentialId, PublicKey publicKey) throws Exception {
        ByteArrayOutputStream authData = new ByteArrayOutputStream();

        // 1. RP ID Hash (32 bytes) - SHA-256 of rpId
        byte[] rpIdHash = PasskeyCrypto.sha256(rpId);
        authData.write(rpIdHash);

        // 2. Flags (1 byte) - 0x5D for registration
        authData.write(FLAGS_REGISTRATION);

        // 3. Sign count (4 bytes, big-endian) - start at 0
        authData.write(new byte[]{0x00, 0x00, 0x00, 0x00});

        // 4. AAGUID (16 bytes) - all zeros for software authenticator
        authData.write(SOFTWARE_AAGUID);

        // 5. Credential ID length (2 bytes, big-endian)
        int credIdLen = credentialId.length;
        authData.write((credIdLen >> 8) & 0xFF);
        authData.write(credIdLen & 0xFF);

        // 6. Credential ID
        authData.write(credentialId);

        // 7. COSE public key (CBOR-encoded)
        byte[] coseKey = encodeCOSEPublicKey(publicKey);
        authData.write(coseKey);

        return authData.toByteArray();
    }

    /**
     * Build authenticator data for passkey assertion.
     * Structure: rpIdHash(32) || flags(1) || signCount(4) = 37 bytes
     */
    public static byte[] buildForAssertion(String rpId) throws Exception {
        ByteArrayOutputStream authData = new ByteArrayOutputStream();

        // 1. RP ID Hash (32 bytes)
        byte[] rpIdHash = PasskeyCrypto.sha256(rpId);
        authData.write(rpIdHash);

        // 2. Flags (1 byte) - 0x1D for assertion
        authData.write(FLAGS_ASSERTION);

        // 3. Sign count (4 bytes, big-endian) - 0
        authData.write(new byte[]{0x00, 0x00, 0x00, 0x00});

        return authData.toByteArray();
    }

    /**
     * Encode P256 public key as COSE_Key (CBOR).
     * Structure: {1: 2, 3: -7, -1: 1, -2: x(32), -3: y(32)}
     *   kty (1): EC2 (2)
     *   alg (3): ES256 (-7)
     *   crv (-1): P-256 (1)
     *   x (-2): x-coordinate (32 bytes)
     *   y (-3): y-coordinate (32 bytes)
     *
     * Key pitfall: ECPublicKey.getW().getAffineX/Y() returns BigInteger which may be
     * less than 32 bytes. Must left-pad to exactly 32 bytes for COSE key encoding.
     */
    private static byte[] encodeCOSEPublicKey(PublicKey publicKey) {
        ECPublicKey ecKey = (ECPublicKey) publicKey;
        byte[] x = padTo32Bytes(ecKey.getW().getAffineX());
        byte[] y = padTo32Bytes(ecKey.getW().getAffineY());

        ByteArrayOutputStream coseKey = new ByteArrayOutputStream();

        // Map with 5 items
        writeBytes(coseKey, CBOREncoder.startMap(5));

        // 1 (kty): 2 (EC2)
        writeBytes(coseKey, CBOREncoder.encodeUInt(1));
        writeBytes(coseKey, CBOREncoder.encodeUInt(2));

        // 3 (alg): -7 (ES256) -> encode as negative: -7 = -1 - 6
        writeBytes(coseKey, CBOREncoder.encodeUInt(3));
        writeBytes(coseKey, CBOREncoder.encodeNegInt(6));

        // -1 (crv): 1 (P-256) -> -1 = -1 - 0
        writeBytes(coseKey, CBOREncoder.encodeNegInt(0));
        writeBytes(coseKey, CBOREncoder.encodeUInt(1));

        // -2 (x): x-coordinate -> -2 = -1 - 1
        writeBytes(coseKey, CBOREncoder.encodeNegInt(1));
        writeBytes(coseKey, CBOREncoder.encodeBytes(x));

        // -3 (y): y-coordinate -> -3 = -1 - 2
        writeBytes(coseKey, CBOREncoder.encodeNegInt(2));
        writeBytes(coseKey, CBOREncoder.encodeBytes(y));

        return coseKey.toByteArray();
    }

    /**
     * Encode attestation object (CBOR map).
     * Structure: {"fmt": "none", "attStmt": {}, "authData": bytes}
     */
    public static byte[] encodeAttestationObject(byte[] authData) {
        ByteArrayOutputStream attestationObject = new ByteArrayOutputStream();

        // Map with 3 items
        writeBytes(attestationObject, CBOREncoder.startMap(3));

        // "fmt": "none"
        writeBytes(attestationObject, CBOREncoder.encodeText("fmt"));
        writeBytes(attestationObject, CBOREncoder.encodeText("none"));

        // "attStmt": {}
        writeBytes(attestationObject, CBOREncoder.encodeText("attStmt"));
        writeBytes(attestationObject, CBOREncoder.startMap(0));

        // "authData": bytes
        writeBytes(attestationObject, CBOREncoder.encodeText("authData"));
        writeBytes(attestationObject, CBOREncoder.encodeBytes(authData));

        return attestationObject.toByteArray();
    }

    /**
     * Build client data JSON for registration.
     * Keys are sorted for consistent output (matching iOS JSONSerialization.sortedKeys).
     */
    public static byte[] buildClientDataJSONForRegistration(byte[] challenge, String origin) {
        return buildClientDataJSON("webauthn.create", challenge, origin);
    }

    /**
     * Build client data JSON for assertion.
     */
    public static byte[] buildClientDataJSONForAssertion(byte[] challenge, String origin) {
        return buildClientDataJSON("webauthn.get", challenge, origin);
    }

    private static byte[] buildClientDataJSON(String type, byte[] challenge, String origin) {
        // Build JSON with sorted keys to match iOS JSONSerialization.sortedKeys
        // Sorted key order: challenge, crossOrigin, origin, type
        String challengeB64 = Base64URLUtils.encode(challenge);
        StringBuilder json = new StringBuilder();
        json.append("{");
        json.append("\"challenge\":\"").append(challengeB64).append("\",");
        json.append("\"crossOrigin\":false,");
        json.append("\"origin\":\"").append(escapeJsonString(origin)).append("\",");
        json.append("\"type\":\"").append(type).append("\"");
        json.append("}");
        return json.toString().getBytes(StandardCharsets.UTF_8);
    }

    /**
     * Pad a BigInteger's byte representation to exactly 32 bytes.
     * BigInteger.toByteArray() may return fewer than 32 bytes (if leading zeros)
     * or 33 bytes (if sign bit needs an extra byte).
     */
    private static byte[] padTo32Bytes(BigInteger value) {
        byte[] raw = value.toByteArray();

        if (raw.length == 32) {
            return raw;
        } else if (raw.length > 32) {
            // BigInteger may prepend a 0x00 byte for positive numbers
            byte[] padded = new byte[32];
            System.arraycopy(raw, raw.length - 32, padded, 0, 32);
            return padded;
        } else {
            // Left-pad with zeros
            byte[] padded = new byte[32];
            System.arraycopy(raw, 0, padded, 32 - raw.length, raw.length);
            return padded;
        }
    }

    private static void writeBytes(ByteArrayOutputStream out, byte[] data) {
        out.write(data, 0, data.length);
    }

    private static String escapeJsonString(String s) {
        return s.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }
}
